Add MoreThuente Linesearcher

vladimir-ch opened the pull request:
I like the names DecreaseFactor and CurvatureFactor. I would vote for using them also in the other Linesearchers.

As I mentioned before, CG is the method that most benefits from this Linesearcher, so I set it as the default. Maybe we should remake the CG tests, there are already many combinations and the Linesearchers add an extra dimension.

PTAL @btracey

btracey commented on linesearcher_test.go:
```diff
@@ -0,0 +1,177 @@
+	g    func(float64) float64
+}
+
+// newPlassmannFunc returns a test function labeled as (5.3) in
```
Shouldn't these be in functions? Even if specialized I think we/someone will find it useful to have an easily accessible set of functions. We could provide a one-D wrapper here or something.

vladimir-ch replied:
Indeed. I will move them there.

vladimir-ch replied:
Done. PTAL @btracey

btracey commented on errors.go:
```diff
@@ -27,6 +27,10 @@ var (
+
+	// ErrMaximumStep signifies that a Linesearcher reached a maximum allowed
+	// step that is not satisfactory.
+	ErrMaximumStep = errors.New("linesearch: reached maximum step")
```
How about this is "ErrLinesearchBound". Some linesearchers have a minimum step, and this may as well cover both?

vladimir-ch replied:
Changed to `ErrLinesearcherBound`. At the moment I will leave it as an ordinary error but I can imagine that storing the step and the bounds could be potentially useful.

btracey commented on functions/functions.go:
```diff
@@ -1627,3 +1627,144 @@ func (Wood) Minima() []Minimum {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+
```
remove newline

vladimir-ch replied:
Done

btracey commented on functions/functions.go:
```diff
@@ -1627,3 +1627,144 @@ func (Wood) Minima() []Minimum {
+	if len(x) != len(grad) {
+		panic("incorrect size of the gradient")
+	}
+
```
remove newline

vladimir-ch replied:
Done.

btracey commented on functions/functions.go:
```diff
@@ -1627,3 +1627,144 @@ func (Wood) Minima() []Minimum {
+// Plassmann implements the univariate function labeled as (5.3) in
+//  More, J.J., and Thuente, D.J.: Line Search Algorithms with Guaranteed Sufficient
+//  Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286–307
+type Plassmann struct {
```
Is there a one sentence why this is interesting?

vladimir-ch replied:
Done. This function is the most interesting of the four.

btracey commented on functions/functions.go:
```diff
@@ -1627,3 +1627,144 @@ func (Wood) Minima() []Minimum {
+//  Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286–307
+// See also
+//  Yanai, H., Ozawa, M., and Kaneko, S.: Interpolation methods in one dimensional
+//  optimization. Computing 27 (1981), 155–163
```
One sentence why this is interesting?

vladimir-ch replied:
Done.

btracey replied:
How about 

````
YanaiOzawaKaneko is a one-dimensional convex function where the values of Beta1 and Beta2 control the curvature around the minimum. Far away from the minimum the function approximates an absolute value function. Near the minimum, the function can either be sharply curved or flat, controlled by the parameter values"

vladimir-ch replied:
Very nice description, thanks! I wanted to mention the absolute value function but couldn't formulate it in a not-too-mathematical way. Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,383 @@
+type MoreThuente struct {
+	// DecreaseFactor is the constant factor in the sufficient decrease
+	// (Armijo) condition.
+	// It must be from the interval [0, 1). The default value is 0.
```
be in the interval

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,383 @@
+	DecreaseFactor float64
+	// CurvatureFactor is the constant factor in the Wolfe conditions. Smaller
+	// values result in a more exact line search.
+	// It must be from the interval [0, 1). If it is zero, it will be defaulted
```
be in the interval

btracey replied:
Maybe we should say "A set value must be in the interval (0, 1). If it is zero..."

vladimir-ch replied:
Done both.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,383 @@
+	// It must be from the interval [0, 1). If it is zero, it will be defaulted
+	// to 0.9.
+	CurvatureFactor float64
+	// StepTolerance is the tolerance for an acceptable step. The line search
```
This is too vague. "StepTolerance sets the minimum acceptable width for the linesearch interval. If the relative interval length is less than this value, ErrLinesearchFailure is returned.

vladimir-ch replied:
Thanks. Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,383 @@
+	// It must be non-negative. If it is zero, it will be defaulted to 1e-10.
+	StepTolerance float64
+
+	// MinimumStep is the minimum step that the linesearcher will take.
```
Should it be the case that StepTolerance is a relative value while Minimu and Maximum are absolute values? Seems odd.

btracey replied:
"Defaults to no minimum (a value of 0)"?

vladimir-ch replied:
I think that Minimum and MaximumStep being absolute is not odd. They are a priori bounds on the step independent of the actual value of the step. Our current optimizers are unconstrained and do not bound the step. But for example Nocedal's LBFGS-B code uses the maximum bound to limit the step length in the given direction and that was also my motivation for including it here.

vladimir-ch replied:
I don't insist on the step tolerance being relative, it's simply what the original code is doing. What would probably make most sense would be to make it an absolute tolerance for the interval length scaled by a norm of the descent direction, something like |step_max - step_min| * ||dir|| < abs_step_tol.

btracey commented on linesearcher_test.go:
```diff
@@ -0,0 +1,129 @@
+				f, g float64
+				step float64
+			)
+		loop:
```
This is a nice test (and not just because of how easy it is under the new invariants!). We may want to think about generalizing it to all the line searchers, i.e.  an "acceptableStep" switch or something.  This would help us make the the behavior similar across the linesearchers, especially in regard to floating point noise.

btracey replied:
This is good for now though until we know how that design should be. Maybe just a "TODO"

vladimir-ch replied:
I don't understand these two comments. Could you please elaborate what you mean by "this test" and "generalize it to all linesearchers"? I already use `func testLinesearcher` to test `Bisection` and `Backtracking` so you probably have something else in mind.

btracey replied:
I see, I misread the test.

vladimir-ch commented:
Addressed comments. I also added a check that `Linesearcher.Init` returns an evaluating operation and changed the invalid operation panic to a regular `t.Errorf`. Sorry for changing the code outside of the comments but it seemed merited to me. PTAL @btracey

btracey commented on functions/functions.go:
```diff
@@ -1627,3 +1627,141 @@ func (Wood) Minima() []Minimum {
+//  Yanai, H., Ozawa, M., and Kaneko, S.: Interpolation methods in one dimensional
+//  optimization. Computing 27 (1981), 155–163
+// This function is convex but different choices of Beta1 and Beta2 lead to
+// functions with different characteristics.
```
characteristics such as?

vladimir-ch replied:
It's hard for me to describe it in one sentence. Please take a look at what I have there now.

btracey commented on linesearcher_test.go:
```diff
@@ -0,0 +1,136 @@
+	}
+}
+
+func testLinesearcher(ls Linesearcher, decrease, curvature float64, strongWolfe bool, t *testing.T) {
```
Could you make t first?

btracey commented:
Sorry, I do understand why Max and Min steps are absolute, but I don't understand why StepTolerance should be relative, especially when the other Step numbers are all absolute.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+	//  psi(step) = f(step) - f(0) - DecreaseFactor * step * f'(0).
+	// When stage is 2, the interval is updated so that it contains a minimizer
+	// of f.
+	stage int
```
Since this only has two values, should it be a boolean?

vladimir-ch replied:
In fact it has three values because I (ab)use the zero value of `stage` to check that `Init` has been called. If you don't like it or if such stringent checking is deemed too pedantic, I can make it a bool.

btracey replied:
No, that's fine.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+	stage int
+
+	step         float64    // Current step.
+	lower, upper float64    // Lower and upper bounds on the next step.
```
On the next step, or just on the step?

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		panic("morethuente: maximum step is not greater than minimum step")
+	}
+	if mt.DecreaseFactor < 0 || mt.DecreaseFactor >= 1 {
+		panic("morethuente: invalid function tolerance")
```
invalid decrease factor would make more sense

vladimir-ch replied:
A leftover from a previous WIP code. Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		panic("morethuente: invalid function tolerance")
+	}
+	if mt.CurvatureFactor <= 0 || mt.CurvatureFactor >= 1 {
+		panic("morethuente: invalid curvature tolerance")
```
same as above

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+	return FuncEvaluation | GradEvaluation, mt.step, nil
+}
+
+// nextStep computes the next safeguarded step and updates the interval that
```
Would you like me to check this algorithm?

vladimir-ch replied:
I tried to factor out the interpolation code but each case differs a bit in such a way that I gave up. Still I feel that it should be possible to make it nicer, so it would be certainly nice if you could check it.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+	// CurvatureFactor is the constant factor in the Wolfe conditions. Smaller
+	// values result in a more exact line search.
+	// A set value must be in the interval (0, 1). If it is zero, it will be
+	// defaulted to 0.9.
```
I think absolute is easier to think about. There's no clear definition of "relative" here (though I know you provided one). Additionally, the tolerance, even absolutely, is already scaled no ||dir||. Our "coordinates" have a step of size ||dir|| equal to 1.

vladimir-ch replied:
> Additionally, the tolerance, even absolutely, is already scaled no ||dir||. Our "coordinates" have a step of size ||dir|| equal to 1.

Yes, that's why I offered the alternative to measure the interval width in real, non-scaled-on-dir units. However, for that we would have to change the `Linesearcher` interface and that does not feel worthy. I don't have a strong argument for a relative tolerance, so I will change it to an absolute tolerance and see what it does with the tests.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		// not been reduced sufficiently after two steps, use bisection to
+		// force its length to zero.
+		width := mt.y - mt.x
+		if math.Abs(width) >= 0.66*mt.width[1] {
```
Can we just make this 2.0/3?

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+	y := mt.y
+	step := mt.step
+
+	gSign := g
```
This is just used as a boolean (g < 0). 

gSign := g > 0
if gx < 0{
    gSign = !gSign
}

vladimir-ch replied:
Did s/gSign/gNeg/ and changed to bool.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		s = math.Max(s, math.Abs(theta))
+		gamma := s * math.Sqrt((theta/s)*(theta/s)-(gx/s)*(g/s))
+		if step < x {
+			gamma = -gamma
```
gamma *= -1?

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		s = math.Max(s, math.Abs(theta))
+		gamma := s * math.Sqrt(math.Max(0, (theta/s)*(theta/s)-(gx/s)*(g/s)))
+		if step > x {
+			gamma = -gamma
```
*= -1

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		if step > x {
+			gamma = -gamma
+		}
+		p := (gamma - g) + theta
```
I don't think we need the parentheses.

vladimir-ch replied:
Yes, that's what I thought too. The Fortran original uses these "unnecessary" parentheses in the interpolation code consistently enough for me to think that there is a purpose (i.e., floating point magic) behind it so I kept them. I will try if removing them has any effect on the tests. At any case, I'll be happy to remove them if you want.

btracey replied:
They do literally nothing though here, I’m pretty sure. The Go spec guarantees they happen in the order as written. What is says:

"Floating-point operations within a single expression are evaluated according to the associativity of the operators.  Explicit parentheses affect the evaluation by overriding the default associativity. In the expression x + (y + z) the addition y + z is performed before adding x. “

So, by default, gamma - g + theta  IS (gamma - g) + theta. I’m not sure what the Fortran rules of this are.


> On Jan 21, 2016, at 7:04 PM, Vladimír Chalupecký <notifications@github.com> wrote:
> 
> In morethuente.go <https://github.com/gonum/optimize/pull/148#discussion_r50494427>:
> 
> > +		}
> > +		bracketed = true
> > +
> > +	case math.Abs(g) < math.Abs(gx):
> > +		// A lower function value, derivatives of the same sign, and the
> > +		// magnitude of the derivative decreases. Extrapolate function values
> > +		// at x and step so that the next step lies between step and y.
> > +
> > +		theta := 3*(fx-f)/(step-x) + gx + g
> > +		s := math.Max(math.Abs(gx), math.Abs(g))
> > +		s = math.Max(s, math.Abs(theta))
> > +		gamma := s * math.Sqrt(math.Max(0, (theta/s)*(theta/s)-(gx/s)*(g/s)))
> > +		if step > x {
> > +			gamma = -gamma
> > +		}
> > +		p := (gamma - g) + theta
> Yes, that's what I thought too. The Fortran original uses these "unnecessary" parentheses in the interpolation code consistently enough for me to think that there is a purpose (i.e., floating point magic) behind it so I kept them. I will try if removing them has any effect on the tests. At any case, I'll be happy to remove them if you want.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/gonum/optimize/pull/148/files#r50494427>.
>

vladimir-ch replied:
One fine day I should read the spec thoroughly. I will remove them.

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+			gamma = -gamma
+		}
+		p := (gamma - g) + theta
+		q := (gamma + (gx - g)) + gamma
```
Same here.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+		s := math.Max(math.Abs(gx), math.Abs(g))
+		s = math.Max(s, math.Abs(theta))
+		gamma := s * math.Sqrt((theta/s)*(theta/s)-(gx/s)*(g/s))
+		if step > x {
```
Same as below, *=-1, no parentheses

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+			}
+			// Modify next if it is close to or beyond y.
+			if step > x {
+				next = math.Min(step+0.66*(y-step), next)
```
2.0/3?

vladimir-ch replied:
Done.

btracey commented on morethuente.go:
```diff
@@ -0,0 +1,384 @@
+			s = math.Max(s, math.Abs(theta))
+			gamma := s * math.Sqrt((theta/s)*(theta/s)-(gy/s)*(g/s))
+			if step > y {
+				gamma = -gamma
```
Same as above.

btracey commented:
LGTM. Thanks.

vladimir-ch commented:
The discussion about the relative step tolerance has been left open-ended but I will merge anyway. I will open an issue and copy the discussion there so that it is together in one place.
